universal-application-tool-0.0.1/app/services/question/QuestionService.java

```diff
@@ -0,0 +1,60 @@
+package services.question;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Locale;
+import java.util.Optional;
+
+public interface QuestionService {
+
+  /**
+   * Creates a new Question Definition. Returns a QuestionDefinition object on success and null on
+   * failure.
+   *
+   * <p>This will fail if: - The path provided already resolves to a QuestionDefinition or Scalar.
+   *
+   * <p>NOTE: This does not update the version.
+   */
+  public abstract Optional<QuestionDefinition> create(QuestionDefinition definition);
+
+  /**
+   * Adds a new translation to an existing question definition. Returns true if the write is
+   * successful.
+   *
+   * <p>The write will fail if: - The path does not resolve to a QuestionDefinition. - A translation
+   * with that Locale already exists for a given question path.
+   *
+   * <p>NOTE: This does not update the version.
+   */
+  public abstract boolean addTranslation(
+      String path, Locale Locale, String questionText, Optional<String> questionHelpText);
+
+  /**
+   * Destructive overwrite of a question at a given path.
+   *
+   * <p>NOTE: This updates the service and question versions.
+   */
+  public abstract QuestionDefinition update(QuestionDefinition definition);
+
+  /** Checks whether a specific path is valid. */
+  public abstract boolean isValid(String pathString);
+
+  /** Gets the question definition for a given path. */
+  public abstract Optional<QuestionDefinition> getQuestionDefinition(String pathString);
+
+  /**
+   * Returns all of the scalar properties for a given path. If the path is invalid this returns
+   * Optional.empty().
+   */
+  public abstract Optional<ImmutableMap<String, ScalarType>> getPathScalars(String pathString);
+
+  /** Gets the type of the node if it exist. Otherwise returns Optional.empty(). */
+  public abstract Optional<PathType> getPathType(String pathString);
+
+  /** Returns all question definitions for this version. */
+  public abstract QuestionDefinition[] getAllQuestions();
+
+  /**
+   * Returns a map of full path to ScalarType for all scalars referenced in the question service.
+   */
+  public abstract ImmutableMap<String, ScalarType> getFullyQualifiedScalars();
```

Comment:
Question: what use case do we have for this? If there isn't one for it yet I recommend not investing in implementing it (same with any others that we can't see an immediate or near-immediate use for).

Reply:
This would essentially give us the empty applicant data model for a given version of the service.

| key | value |
| -------- | -------- |
| applicant.name.first | string |
| applicant.name.last | string |

Reply:
I'm not sure we have anything in scope for M1 or M2 that needs that. Also (and unfortunately), repeated questions will mean we can't generate a complete empty applicant data model.

Reply:
Achievement unlocked: Scope reduced!